### PR TITLE
Fix an infinite loop when using `#fmt: on/off` in the middle of an expression or code block

### LINKED ISSUE #2015: Hang when parsing `# fmt: on` in the middle of an expression
**Describe the bug**
`black` hangs when parsing `# fmt: on` in the middle of an expression

**To Reproduce**
1. Take this file
```py
from subprocess import run

path = ["."]
run(
    # fmt: off
    [
        "ls",
        "-la",
    ]
    # fmt: on
    + path,
    check=True,
)
```
2. Run _Black_ on it with these arguments `--target-version py38`
3. Observe that black never exits.

**Expected behavior**
Either black should format the file fine (19.10b0 does not have trouble here) or it should error out and tell me it failed to format.

**Environment (please complete the following information):**

- Version: [e.g. master] 20.8b1
- OS and Python version: [e.g. Linux/Python 3.7.4rc1] Linux/Python 3.8.5

**Does this bug also happen on master?** To answer this, you have two options:

1. Use the online formatter at https://black.now.sh/?version=master, which will use the
   latest master branch.
2. Or run _Black_ on your machine:
   - create a new virtualenv (make sure it's the same Python version);
   - clone this repository;
   - run `pip install -e .`;
   - make sure it's sane by running `python -m unittest`; and
   - run `black` like you did last time.

Yes, it is happening on the online formatter as well.

**Additional context**
If I remove `check=True,` from my example, it works fine. If I move `# fmt: on` after `+ path,` it works fine.

We've modified our code to put `# fmt: on` in a more reasonable location, but still wanted to report it as I wouldn't expect a hang.

### LINKED ISSUE #3026: `yapf: enable` comments inside expression cause infinite hang
**Describe the bug**

If you have `yapf: enable` comments inside an expression, you can cause black to hang:

**To Reproduce**

with `file.py` as:
```python
def test_func():
    # yapf: disable
    if a:
        return True
    # yapf: enable
    elif b:
        return True

    return False
```

```sh
$ black file.py
```

It hangs forever.

**Expected behavior**

It should not hang.

At the very least it should give an error.

Preferably it should ignore those comments and not break anything.

**Environment**

- Black's version: black, 22.3.0 (compiled: yes)
- OS and Python version: Python 3.8 & 3.10, mac OS 12.3

Also happens on the https://black.vercel.app/

### LINKED ISSUE #3129: Black hangs when using fmt:on/off improperly
**Description**

The following code fails:

```python
# -*- coding: utf-8 -*-

setup(
    entry_points={
        # fmt: off
        "console_scripts": [
            "foo-bar"
            "=foo.bar.:main",
        # fmt: on
        ]
    },
)
```

The following code works:

```python
# -*- coding: utf-8 -*-

setup(
    entry_points={
        # fmt: off
        "console_scripts": [
            "foo-bar"
            "=foo.bar.:main",
        ]
    # fmt: on
    },
)
```

While running it with these arguments:

```sh
$ black file.py
```

For the first code, there is no error: it just hangs indefinitely. It is also impossible to "CTRL + C" the command.

**Expected behavior**

Either:
- a fail ("You are using fmt wrong")
- or change the "fmt" functionality to make the first snippet above work.

**Environment**

- Black's version: 22.3.0
- OS and Python version: Ubuntu 20.04 / 21.04

### PR DESCRIPTION
**Pull request by yilei**
### Description

Fixes an infinite loop when using `#fmt: on/off` in the middle of an expression or code block.

This:

- resolves #3129
- resolves #3026
- resolves #2015

### Checklist - did you ...

- [x] Add a CHANGELOG entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Review by ichard26 — COMMENTED** [review]
The `# fmt: off/on/skip` might be the most hacky and impossibly-to-maintain complex feature of Black. Writing any sort of robust code here is challenging. This is definitely a step forward, but there's a wart that I'm not sure how to resolve.

My suggestion is to just fix the weird formatting issue as it involves a (IMO) valid use of `# fmt: off/on` and let the other usages break since they don't use it properly`. OTOH, there might be cases with valid usages of `# fmt: off/on` that break by fixing the formatting issue with #2015's example ...

**Comment by yilei** [comment]
> `# fmt: off/on/skip` might be the most hacky, impossibly-hard-to-maintain, and complex feature of Black

The `STANDALONE_COMMENT` a neat trick though :)

Thanks for the review and suggestion!

I updated the code to your suggested "if child is a Leaf (so doesn't contain children) and contains a # fmt: on, then return", so the weird `+ path,` formatting is fixed.

I also added another hack to include this child Leaf if this is a closing bracket (it's an invalid use), so it doesn't fail for the first example.

Please take another look.

File: tests/data/simple_cases/fmtonoff5.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# Regression test for https://github.com/psf/black/issues/3129.
+run(
+    # fmt: off
+    [
+        "ls",
+        "-la",
+    ]
+    # fmt: on
+    + path
+    ,
```
[the remark this answers is not in the document]
Comment (yilei) [reply tests/data/simple_cases/fmtonoff5.py]: Updated so this is now fixed.

**Review by ichard26 — APPROVED** [review]
SGTM. Thank you for tackling this bug! 

P.S. you need to add the closing keyword before each issue you'd like to close automagically 
image

**Comment by yilei** [comment]
thanks for the tip! i updated this PR (and #3148)'s description to use the correct syntax.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**